Add reading of rating tag for MP3, MP4, and FLAC

LindyBalboa opened the pull request:
So this is partly addressed in the open PR #924 but it seems to have stagnated so I have decided to jump on it. Current state is that for mp3 the rating is taken from a POPM tag with email "mixxx@mixxx.org" or the first available if there is none from mixxx. MP4 and FLAC both use a plain 0-100 integer scale. 

Values are then parsed into a 0-5 value before being added to the metadata object. For POPM, the same standard as Windows Media Player/Explorer is followed. 

To do:
-add support for APE tags I'm not familiar with it and don't use it. I can't seem to find the key for the rating tag. Does anyone know this off hand?
-add writing from database to file. Need some discussion here. As MP4, FLAC, and APE do no use POPM, there isn't the email feature. So I feel it would actually make little sense to use a designated email for mixxx. Some users claim that they need different rating option for different applications/people, but honestly I use a mixed library; format isn't so important to me. Like I said, discussion needed.
-add the option to not overwrite existing ratings in the database
-add the option to turn writing the rating to the file tag on or off

I'm not quite sure where in the code the writing part would be controlled. If someone could point me in the right direction I'm sure I can figure it out, but if someone wants to contribute that, that would be great too.

LindyBalboa committed:
Add reading of rating tag for MP3, MP4, and FLAC

uklotzde commented:
Please replace all tabs with 4 spaces:
(https://www.mixxx.org/wiki/doku.php/coding_guidelines)

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -232,6 +232,71 @@ bool parseBpm(TrackMetadata* pTrackMetadata, QString sBpm) {
     return isBpmValid;
 }
 
+void parseRating(TrackMetadata *pTrackMetadata, QString sRating, FileType filetype = FileType::UNKNOWN) {
```
The context (i.e. file/tag type) is known at compile time when this function is invoked. Please split it into 3 functions.

uklotzde replied:
You can omit the pTrackMetadata argument if you simply return the parsed rating and let the caller decide what to do with this result. Currently this function does two different things: parsing + setting. But the name is only "parse".

uklotzde replied:
Always prefer to write functions with a result and no side effects over functions with no result but with side effects.

uklotzde commented:
Tag writing: Simply add the conversion in the opposite direction from iRating into tag values to the specialized write functions in trackmetadatataglib.cpp.

Writing tags into files is currently disabled:
(https://github.com/mixxxdj/mixxx/pull/728)
With this PR all tag values including the rating could be written into files.

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -967,6 +1032,20 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
+		rating = QString::number(dynamic_cast<TagLib::ID3v2::PopularimeterFrame*>(popmRatingFrameList.front())->rating());
+		for (TagLib::ID3v2::FrameList::ConstIterator it(popmRatingFrameList.begin());
+			it != popmRatingFrameList.end(); it++){
+		        auto popmFrame = dynamic_cast<TagLib::ID3v2::PopularimeterFrame*>(*it);
```
The result of a dynamic_cast should always be checked! See various examples in this file

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -967,6 +1032,20 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
+	QString rating;
+    const TagLib::ID3v2::FrameList popmRatingFrameList(tag.frameListMap()["POPM"]);
+	if (!popmRatingFrameList.isEmpty()) {
+		rating = QString::number(dynamic_cast<TagLib::ID3v2::PopularimeterFrame*>(popmRatingFrameList.front())->rating());
```
This code duplicates part of the loop body. Initialize the rating with -1 and inside the loop body check if either of these conditions is met: (rating == -1) || (email == ...)

LindyBalboa replied:
The point for the duplicaton was to take the first rating found and then check all of them if there is a mixxx specific frame. Since this is only truly a problem for MP3, an alternative would be to forget having the mixxx specific email and just reading writing the 1st frame. 

That is really a decision that should be made by people more familiar with the trajectory of the project.

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -967,6 +1032,20 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
+			it != popmRatingFrameList.end(); it++){
+		        auto popmFrame = dynamic_cast<TagLib::ID3v2::PopularimeterFrame*>(*it);
+			    if (toQString(popmFrame->email()) == "mixxx@mixxx.org") {
+				    rating = QString::number(popmFrame->rating());
```
You should exit the loop after you found the first rating with a matching email.

uklotzde commented on src/track/trackmetadata.h:
```diff
@@ -140,6 +140,14 @@ class TrackMetadata {
         m_bpm.resetValue();
     }
 
+	// rating
```
The comment "// rating" does not add any value.

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -967,6 +1032,20 @@ void readTrackMetadataFromID3v2Tag(TrackMetadata* pTrackMetadata,
+		for (TagLib::ID3v2::FrameList::ConstIterator it(popmRatingFrameList.begin());
+			it != popmRatingFrameList.end(); it++){
+		        auto popmFrame = dynamic_cast<TagLib::ID3v2::PopularimeterFrame*>(*it);
+			    if (toQString(popmFrame->email()) == "mixxx@mixxx.org") {
```
The e-mail address should be stored in a named constant. Simply define one in the anonymous namespace in the .cpp file.

uklotzde commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -1256,6 +1340,11 @@ void readTrackMetadataFromMP4Tag(TrackMetadata* pTrackMetadata, const TagLib::MP
 
+	QString rating;
+	if (readMP4Atom(tag, "rate", &rating)) {
+	    parseRating(pTrackMetadata, rating, FileType::MP4);
```
As already mentioned, the FileType is always known at compile time and does not need to passed as an additional argument.

LindyBalboa commented:
All good points. I have been on vacation for a bit and will be getting back into work mode over the next few days but will jump back on this when I get back up to speed!

LindyBalboa committed:
Split parseRating remove side effects check dynamic cast

parseRating has been split into separate functions for MP3, FLAC, and
MP4 filetypes since  they are already known.
parse*TYPE*Rating now only returns the value, without side effects.
Now check when casting a generic TagLib frame to a popularimeter frame.

LindyBalboa committed:
Convert ratings when reading instead when parsing

For MP4 and Flac, the convenience functions to read tag items
assumes all fields are strings. Previously conversion was done
in the parsing functions, it has been switched to write after
reading, before the parsers are called.

LindyBalboa committed:
Change id3 reader to only take the first POPM tag found

LindyBalboa commented:
I think this should have a default off operation and only be turned on by choice. It has the potential to overwrite the current database-only ratings. Once writing metadata is implemented, it should also have the option "write all nonzero ratings to file before reading ratings from files".

Thoughts?

uklotzde commented:
From a usability viewpoint this decision is far too complex for a user 
without background knowledge. It is the application developer's 
responsibility to properly design workflows that work as expected and 
doesn't expose any implementation details.

Make a decision table and write the desired outcome with any possible side 
effects and how they can be solved. Then decide how this can be implemented 
in code and database schema without asking the user to make any decisions.

Example: In previous versions Mixxx didn't read the total number of tracks 
(track total). For some formats this number is written together with the 
number of tracks in the same field. We decided to initialize the new 
database field with an (invalid) default value '//'. Whenever we encounter 
such a track we re-read track number + total (ignoring any other tags) and 
replace the default value in the Mixxx database with the actual values, e.g. 
'3/11'. With this strategy all missing values will be read eventually and 
(important!) before overwriting any existing tags. Otherwise the track total 
tag in the track's file might get lost when writing tags.

On 19.03.2017 21:42, Conner Phillips wrote:
>
> I think this should have a default off operation and only be turned on by 
> choice. It has the potential to overwrite the current database-only 
> ratings. Once writing metadata is implemented, it should also have the 
> option "write all nonzero ratings to file before reading ratings from files".
>
> Thoughts?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1089#issuecomment-287645916>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-akV94s56c2884Kl6q13XEwkOQYD-ks5rnZM9gaJpZM4LUxHN>.
>

illuusio commented:
What is this current situation? Should this work and is there test case for this implemented so I can run this on test pack?

LindyBalboa commented:
This does work. There is no testing implemented. It is basically bare bones minimum working example as of yet. I've really been tied up with school lately and haven't had time to sit down with this. However, I am trying to keep it up to date with master. 

I am completely open to suggestions on how to continue with this.

illuusio commented:
Little testing advise would be nice. How to add these in MP3, M4A or OGG so I can test this.

LindyBalboa commented:
I'm not sure what you are asking? This is built directly in to the standard "read tags from file" operations. Nothing else is required at the moment beyond scanning the library.

Currently Mixxx does not write any tags to file, it only stores them in the database. As @uklotzde wrote, #728 is concerned with mix writing ratings. So this will only pick up ratings generated by other programs. I'm not sure if iTunes is writing ratings to file yet, or if it is all still stored in their XML file. I personally have used MediaMonkey while on Windows and have written my ratings using that up until now. 

So basically, rate some files in another program that we know works. Amarok or Banshee maybe for linux? WMP on Windows I think. Then scan them into Mixxx's library and see if they result is as expected. Half star ratings from programs like MediaMonkey should be rounded up to the next full star.

illuusio commented:
Sorry not being clear about my thoughts! I have to investigate which app I can use to test this. I haven't yet read code very hard but on mp3 it would read ID3v2 **POPM** tag and on Ogg Vorbis it would be **RATING** comment and m4a **RATE** tag. Just wondering which tags to use for testing this out.

uklotzde commented:
You might enrich our test files in src/test/id3-test-data with those tags 
and add the corresponding test cases.

On 07.04.2017 14:18, Tuukka Pasanen wrote:
>
> Sorry not being clear about my thoughts! I have to investigate which app I 
> can use to test this. I haven't yet read code very hard but on mp3 it 
> would read ID3v2 *POPM* tag and on Ogg Vorbis it would be *RATING* comment 
> and m4a *RATE* tag. Just wondering which tags to use for testing this out.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1089#issuecomment-292521501>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-akoEYD0WgN4oT2Zcji_CAvtDfyNpks5rtil9gaJpZM4LUxHN>.
>

LindyBalboa commented:
@uklotzde I can definitely add some tests there. A few questions. 

1) The folder is id3, but this is more general. Should I just place the files in there as is, or should we rename the folder to be more general like test-tag-data?

2) For ratings there are a wide range of cases-- basically the starting and ending points of ranges which become encapsulated in a singular star rating. That would suggest upwards of #ofFormats*#ofStars*2. In this case that comes out to 30. That seems like a lot of test files to include. Any suggestion on limiting the number of cases, but still providing a reasonable amount of testing?

illuusio commented:
First there should be test files with and without stars and maybe different files have different staring or something like that. Added thirty test files sounds awful lot to me.

illuusio commented:
Ok. i'll give this a shot and check how this is working

wrkb commented:
Just to let you guys know: Your work is very appreciated (i do 👍 ) and there are enough scenarios out there, in which the rating is nescessary/helpful.

In my case, I have a bigger collection of files (5000 or so) from http://remix.kwed.org with complex sounds, that I have to hear. So i need ratings to filter out, what needs a second listening later and i can mark some hymns (5 stars), which are always possible for a dance crowd. 3 stars is good, but only for private listening and rarely for dj setz. And so son.

I deeply hope, your contribution will make in the mixxx code!

Thx and greetings!

Be-ing commented:
@LindyBalboa merge conflicts have developed since #728 was merged.

Be-ing commented:
This pull request has not been updated in over a year and merge conflicts have developed, so I am closing it. @LindyBalboa if you fix the merge conflicts and would like to continue working on this, please reopen this PR.

joaonunatings commented:
Is this still being developed @LindyBalboa? I can take over this PR if possible

ronso0 commented:
The last update before your post was that Be closed this because there has not been any progress for over a year.
Since then Be and uklotz left the dev team.

If you want to take over feel free to do so. Though, in order to avoid repetition (and maybe frustration), I recommend you read the entire conversation here, and probably also in #9477 and #924.

joaonunatings commented:
@ronso0 thank you for the quick reply!

I already went through those links (that's how I got to this PR), so I'll probably just create a new issue as a tabula rasa for this problem and mention implementation details, issue/PR mentions and additional info.

ronso0 commented:
Hej @CaqKa 
These are too many questions for a ratings PR. In the future, try to file separate issues / feature requests so they can be tracked easily and the conversations are focused.

Let's break it down:
1. Ratings, plan or discussion about supporting rating import/export in the future
Please check the related feature requests to get an idea what's up / possible / desired / to do
https://github.com/mixxxdj/mixxx/issues?q=is%3Aissue%20state%3Aopen%20rating+import
If you're able, you could also write a helper tool (c++, python) that does a one-time import: reads ratings from individual tracks or a library (iTunes xml) and import them into the `rating` column in the `library` table of mixxxdb.sqlite. Track file locations are stored in the `track_locations` table.
2. Library paths in Flatpak
For issue with the Flatpak build I suggest you file a report at Mixxx' Flathub repo
https://github.com/flathub/org.mixxx.Mixxx/issues
3. Music library portability guidance
For migrating settings and database we have a guide in the manual
https://manual.mixxx.org/latest/chapters/advanced_topics#migrate-your-mixxx-library-and-settings-to-a-new-computer
